### PR TITLE
Add a module to parse XSPEC model.dat and helper scripts

### PR DESCRIPTION
**Pull request by DougBurke**
# Summary

Add support for parsing XSPEC model.dat files (useful for updating the XSPEC module or if you want to use XSPEC local models).

# Details

The `sherpa.astro.utils.xspec` module is added to help handle XSPEC model routines: primarily parsing a model.dat file but also some helper routines to create the code we need to wrap these models.

Notes about the helper scripts has been added to the developer documentation at https://sherpa--1260.org.readthedocs.build/en/1260/developer/index.html#checking-against-a-previous-xspec-version

This code is primarily useful for

a) updating the XSPEC module to a new XSPEC release
   - `scripts/check_xspec_update.py`
     can be used to see what changes are needed
   - `scripts/add_xspec_model.py`
     will print out the basic code needed to support a new XSPEC
     model

   The idea is to automate the code generation but there are still a number of changes that need to be done manually

b) for users who want to use XSPEC user models with Sherpa. This functionality is currently provided by the CIAO contributed code `create_xspec_user_model` from which much of this commit was taken (i.e. it will simplify that script).

# Examples

This requires a Sherpa build of this PR with XSPEC support built in and the examples were run after #1261 was merged (unless mentioned).

## Issues with the model.dat file

We can see issues with models in the XSPEC model.dat that will need some decision on what to do: using HEASOFT 6.29 (XSPEC 12.12.0) we see

```
% python scripts/check_xspec_update.py ~/local/heasoft-6.29/spectral/manager/model.dat | grep WARN
WARNING: model=vvwdem re-uses parameter name p
WARNING: model=ismabs re-uses parameter name sii
WARNING: model=ismabs re-uses parameter name siii
```

This tells us that the vvwdem has two "p" parameters and ismabs has four problem parameters - this is because our parameters are case insensitive which means that p and P end up mapping to the same parameter (and for ismabs it is really fun as you have SII and SiI and SIII and SiII). This warning tells you there's a problem you will need to look at (we have a workaround for ismabs and vvwdem is added in 12.12.0 as discussed in #1223

## Do our model parameter values differ to XSPEC?

Using the HEASOFT 6.28 model.dat file before #1261 was merged we see a number of issues:

```
% ./scripts/check_xspec_update.py ~/local/heasoft-6.28/spectral/manager/model.dat
WARNING: model=ismabs re-uses parameter name sii
WARNING: model=ismabs re-uses parameter name siii
Model agnslim
  [ 1] par logmdot default: -1.0 -> 1.0
  [ 2] par astar softmin: -1.0 -> 0.0
  [ 3] par astar hardmin: -1.0 -> 0.0

Model apec
  [ 1] par kT hardmin: 0.0 -> 0.008
  [ 2] par kT hardmax: 3.4028234663852886e+38 -> 64.0
  [ 3] par Abundanc hardmax: 3.4028234663852886e+38 -> 5.0
  [ 4] par 3 name: redshift -> Redshift
  [ 5] par redshift hardmax: 3.4028234663852886e+38 -> 10.0

Model bapec
  [ 1] par kT hardmin: 0.0 -> 0.008
  [ 2] par kT hardmax: 3.4028234663852886e+38 -> 64.0
  [ 3] par Abundanc hardmax: 3.4028234663852886e+38 -> 5.0
  [ 4] par Redshift hardmax: 3.4028234663852886e+38 -> 10.0
  [ 5] par Velocity hardmax: 3.4028234663852886e+38 -> 1000000.0

Model bbody
  [ 1] par kT hardmin: 0.0 -> 0.0001
  [ 2] par kT hardmax: 3.4028234663852886e+38 -> 200.0
  [ 3] par norm units: L39 / (D10)**2 ->

Model bbodyrad
  [ 1] par kT hardmin: 0.0 -> 0.0001
  [ 2] par kT hardmax: 3.4028234663852886e+38 -> 200.0

...
```

Now that #1261 has been merged we see less issues - many of these warnings could be removed but it's not worth the complexity.

```
%  python scripts/check_xspec_update.py ~/local/heasoft-6.28/spectral/manager/model.dat
WARNING: model=ismabs re-uses parameter name sii
WARNING: model=ismabs re-uses parameter name siii
Model apec
  [ 1] par 3 name: redshift -> Redshift

Model bexrav
  [ 1] par 9 name: redshift -> Redshift

Model bexriv
  [ 1] par 6 name: redshift -> Redshift

Model brnei
  [ 1] par 5 name: redshift -> Redshift

Model c6mekl
  [ 1] par 9 name: redshift -> Redshift
  [ 2] par switch softmin: -3.4028234663852886e+38 -> None
  [ 3] par switch softmax: 3.4028234663852886e+38 -> None
  [ 4] par switch hardmin: -3.4028234663852886e+38 -> None
  [ 5] par switch hardmax: 3.4028234663852886e+38 -> None

...
```

Note that the idea is to point out things that need looking at rather than forcing equality of the two representations (e.g. the redshift/Redshift name differences).

## Are we missing things in XSPEC 12.12.0

This assumes #1223 has not been merged.  Note that XSPEC 12.12.0 was released with HEASOFT 6.29 and I do the `diff` trick below so that the output is not swamped with the parameter differences that #1261 will hopefully clean up.

For reference heasoft 6.28 includes XSPEC 12.11.1.

```
% ./scripts/check_xspec_update.py ~/local/heasoft-6.28/spectral/manager/model.dat > a
% ./scripts/check_xspec_update.py ~/local/heasoft-6.29/spectral/manager/model.dat > b
% diff a b
0a1
> WARNING: model=vvwdem re-uses parameter name p
72a74,75
> We do not support grbjet (Add; xsgrbjet)
>
212,217c215,218
<   [ 2] par redshift softmin: 0.0 -> -0.999
<   [ 3] par redshift hardmin: 0.0 -> -0.999
<   [ 4] par switch softmin: -3.4028234663852886e+38 -> None
<   [ 5] par switch softmax: 3.4028234663852886e+38 -> None
<   [ 6] par switch hardmin: -3.4028234663852886e+38 -> None
<   [ 7] par switch hardmax: 3.4028234663852886e+38 -> None
---
>   [ 2] par switch softmin: -3.4028234663852886e+38 -> None
>   [ 3] par switch softmax: 3.4028234663852886e+38 -> None
>   [ 4] par switch hardmin: -3.4028234663852886e+38 -> None
>   [ 5] par switch hardmax: 3.4028234663852886e+38 -> None
254a256,261
> We do not support vvwdem (Add; vvwDem)
>
> We do not support vwdem (Add; vwDem)
>
> We do not support wdem (Add; wDem)
>
385a393,394
>
> We do not support zxipab (Mul; zxipab)
```

The important lines here are the `We do not support ...`, which point to models that we may want (almost certainly) to add (and #1223 does add them) as well as the WARNING line which tells us we need to think about how we handle the vvwdem model.

Note that there are models we do not support - e.g. the smaug model (because it needs functionality from #1239 **and** then some serious thinking) and the pileup model because we don't have an interface for the "Acn" model type:

```
% grep 'We do not' a
We do not support smaug (Add; xsmaug)
We do not support pileup (Acn; pileup)
```

## What do I need to do to add a model?

From above I know we need to add the wdem model, so we can call `add_xspec_model.py` with the model.dat file and model name and get the C++ and Python code to start the conversion. It's not perfect:

- the C++ code needs some work to add in the necessary code as it can be scattered in the defines and wrapper sections
- the Python code is missing documentation and also any necessary @version_at_least decorators and setting __function__ rather than _calc (which will be needed for optional models)

but it should allow you to at least build the models. This script also does not tell you what needs changing in the test suites (e.g. the number of expected models).

```
% ./scripts/add_xspec_model.py ~/local/heasoft-6.29/spectral/manager/model.dat wdem
# C++ code for sherpa/astro/xspec/src/_xspec.cc

// Defines

extern "C" {

}

// Wrapper

static PyMethodDef Wrappers[] = {
  XSPECMODELFCT_C_NORM(C_wDem, 8),
  { NULL, NULL, 0, NULL }
}


# Python code for sherpa/astro/xspec/__init__.py


class XSwdem(XSAdditiveModel):
    """XSPEC AdditiveModel: wdem

    Parameters
    ----------
    Tmax
    beta
    p
    nH
    abundanc
    Redshift
    switch
    norm

    """
    _calc = _models.C_wDem

    def __init__(self, name='wdem'):
        self.Tmax = Parameter(name, 'Tmax', 1.0, min=0.01, max=10.0, hard_min=0.01, hard_max=20.0, units='keV')
        self.beta = Parameter(name, 'beta', 0.1, min=0.01, max=1.0, hard_min=0.01, hard_max=1.0)
        self.p = Parameter(name, 'p', 0.25, min=-1.0, max=10.0, hard_min=-1.0, hard_max=10.0)
        self.nH = Parameter(name, 'nH', 1.0, min=1e-05, max=1e+19, hard_min=1e-06, hard_max=1e+20, frozen=True, units='cm^-3')
        self.abundanc = Parameter(name, 'abundanc', 1.0, min=0.0, max=10.0, hard_min=0.0, hard_max=10.0, frozen=True)
        self.Redshift = Parameter(name, 'Redshift', 0.0, min=-0.999, max=10.0, hard_min=-0.999, hard_max=10.0, frozen=True)
        self.switch = Parameter(name, '_switch', 2, alwaysfrozen=True)
        self.norm = Parameter(name, 'norm', 1.0, min=0.0, max=1e+24, hard_min=0.0, hard_max=1e+24)
        XSAdditiveModel.__init__(self, name, (self.Tmax,self.beta,self.p,self.nH,self.abundanc,self.Redshift,self.switch,self.norm))
```

### CODE REVIEW

File: sherpa/astro/utils/xspec.py [hunk 1]
```diff
@@ -0,0 +1,1064 @@
...
+                                    softmin, softmax,
+                                    hardmin, hardmax, delta)
+
+
+def add_xs_prefix(inval):
+    """Returns XS prepended to inval"""
+    return f"XS{inval}"
+
+
+def parse_xspec_model_file(modelfile, namefunc=None):
```
[the remark this answers is not in the document]
Comment (DougBurke) [reply sherpa/astro/utils/xspec.py:685]: I prefer the docstring you get:

```python
>>> def foo():
...     pass
...
>>> def bar(bob=foo):
...     pass
...
>>> def bax(bob=None):
...     pass
...
```

so

```python
>>> help(bax)
Help on function bax in module __main__:

bax(bob=None)
```

vs

```python
>>> help(bar)
Help on function bar in module __main__:

bar(bob=<function foo at 0x7f33d4191050>)
```

File: sherpa/astro/utils/xspec.py [hunk 2]
```diff
@@ -0,0 +1,1064 @@
...
+                                                                   hdrline))
+
+    # safety check on the parameter names.
+    #
+    pnames = [(par.name.lower(), par.name) for par in pars]
+    lnames = set(pnames)
+    if len(lnames) != len(pars):
+        d = defaultdict(list)
+        for (k, v) in pnames:
+            d[k].append(v)
```
[the remark this answers is not in the document]
Comment (DougBurke) [reply sherpa/astro/utils/xspec.py:475]: I keep on forgetting about `collections`

File: sherpa/astro/utils/xspec.py [hunk 3]
```diff
@@ -0,0 +1,1064 @@
...
+
+        multiple = [vs for (k, vs) in d.items() if len(vs) > 1]
```
[the remark this answers is not in the document]
Comment (DougBurke) [reply sherpa/astro/utils/xspec.py:477]: I think there was actually an error somewhere in my code. Changing to use Counter ended up showing this so a useful update.

**Comment by DougBurke** [comment]
I've rebased to get recent changes and addressed some of the changes Moritz mentioned (missing some of them).

One useful addition is a test suite. I'm interested to check the coverage.

File: scripts/add_xspec_model.py [hunk 4]
```diff
@@ -0,0 +1,81 @@
...
+
+    Parameters
+    ----------
+    models : list of ModelDefinition
+    modelname : str
+        XSPEC model name (as given in the model.dat file)
+
+    """
+
+    model = None
```
[the remark this answers is not in the document]
Comment (DougBurke) [reply scripts/add_xspec_model.py:53]: Ta

[the remark this answers is not in the document]
Comment (hamogu) [reply scripts/add_xspec_model.py:53]: I like you. Tatata.

File: scripts/check_xspec_update.py [hunk 5]
```diff
@@ -0,0 +1,179 @@
...
+        print(f"Model {mdl.name}")
+        for i, report in enumerate(reports, 1):
+            print(f"  [{i:2d}] {report}")
+
+        print("")
+
+    # Check to see if we have lost any models.
+    #
+    for name in dir(xspec):
+        if not name.startswith('XS') or name.endswith('Model') or name.endswith('Kernel'):
```
[the remark this answers is not in the document]
Comment (DougBurke) [reply scripts/check_xspec_update.py:145]: Unfortnately not - we want to reject XSTableModel (for instance) with this check but it is a model, so it's easiest just to use a name check.

File: sherpa/astro/utils/xspec.py [hunk 6]
```diff
@@ -0,0 +1,1064 @@
...
+#  with this program; if not, write to the Free Software Foundation, Inc.,
+#  51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
+#
+
+"""Parser for XSPEC model files.
+
+The XSPEC library [1]_ uses ASCII files to define models [2]_, and it
+can be useful to be able to read these files either to update the
+Sherpa code to support a new XSPEC release [3]_ or to build support
+local for a user model.
```
[the remark this answers is not in the document]
Comment (DougBurke) [reply sherpa/astro/utils/xspec.py:26]: Text improved now.

**Comment by DougBurke** [comment]
Re-rebased as I'd missed a few comments from @hamogu which turned out to be useful to address

**Comment by DougBurke** [comment]
@hamogu - you may want to have a look at the changes as they are slightly-more than I originally expected.

**Review by hamogu — APPROVED** [review]
It occurs to me now that the output C++ and Python code for this module might be cleaner if using string templating (using https://docs.python.org/3/library/string.html#template-strings or even a full templating library such as jinja2, which developers need as a dependence of sphinx anyway) instead of adding lines by line with `out +=`  and `append`. That way, the basic structure would be defined in one human-readable string instead of being scattered over several lines of code.

However, what @DougBurke has written here works and I don't want too spend too much time on what should be a simple helper script. I'm leaving the comment here to keep in mind if we develop this further to make the addition of new models even more automatic. Of course, if @DougBurke feels like developing this idea further right now, I won't object; but I'm approving as is.

**Pull request merged**